packages/core/src/sync-realtime/service.ts

```diff
@@ -202,444 +219,506 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
 
-        // Find any logs matching log filters or child contract filters.
-        // NOTE: It might make sense to just insert all logs rather than introduce
-        // a potentially slow DB operation here. It's a tradeoff between sync
-        // latency and database growth.
-        const factoryLogFilters = await Promise.all(
-          this.sources.filter(sourceIsFactory).map(async (factory) => {
-            const iterator = this.syncStore.getFactoryChildAddresses({
-              chainId: this.network.chainId,
-              factory: factory.criteria,
-              upToBlockNumber: hexToBigInt(block.number!),
-            });
-            const childContractAddresses: Hex[] = [];
-            for await (const batch of iterator) {
-              childContractAddresses.push(...batch);
-            }
-            return {
-              address: childContractAddresses,
-              topics: factory.criteria.topics,
-            };
-          }),
-        );
+    // Probability of no logs in the range of blocks
+    const pNoLogs = (1 - pLog) ** numBlocks;
+    const costTraverse = 16 * numBlocks + 75 * (1 - pNoLogs);
 
-        matchedLogs = filterLogs({
-          logs,
-          logFilters: [
-            ...this.sources.filter(sourceIsLogFilter).map((l) => l.criteria),
-            ...factoryLogFilters,
-          ],
-        });
-      }
+    if (costBatch > costTraverse) return "traverse";
+    else return "batch";
+  };
 
-      const matchedLogCount = matchedLogs.length;
-      const matchedLogCountText =
-        matchedLogCount === 1
-          ? "1 matched log"
-          : `${matchedLogCount} matched logs`;
+  private syncTraverse = async (newBlock: RealtimeBlock) => {
+    const newBlockNumber = hexToNumber(newBlock.number);
+    const newBlockTimestamp = hexToNumber(newBlock.timestamp);
 
-      this.common.logger.debug({
-        service: "realtime",
-        msg: `Got ${logs.length} total and ${matchedLogCountText} in block ${newBlock.number} (network=${this.network.name})`,
-      });
+    if (newBlockNumber > this.mostRecentBlock.number) {
+      const missingBlockRange = range(
+        this.mostRecentBlock.number + 1,
+        newBlockNumber,
+      );
+      const newBlocks = await Promise.all(
+        missingBlockRange.map(this._eth_getBlockByNumber),
+      );
+      newBlocks.push(newBlock);
 
-      // If there are indeed any matched logs, insert them into the store.
-      if (matchedLogCount > 0) {
-        // Filter transactions down to those that are required by the matched logs.
-        const requiredTransactionHashes = new Set(
-          matchedLogs.map((l) => l.transactionHash),
-        );
-        const filteredTransactions =
-          newBlockWithTransactions.transactions.filter((t) =>
-            requiredTransactionHashes.has(t.hash),
+      if (newBlocks[0].parentHash !== this.mostRecentBlock.hash) {
+        await this.reorgTraverse(newBlocks[0], hexToNumber(newBlock.number));
```

Comment:
Do you need to return here after this?

Reply:
Haven't figured it out yet, not sure you do because right above you just fetched a bunch of new blocks and they might still be useful.